numpy/_array_api/_array_object.py

```diff
@@ -0,0 +1,891 @@
+
+        NumPy deviates from the spec type promotion rules in cases where one
+        argument is 0-dimensional and the other is not. For example:
+
+        >>> import numpy as np
+        >>> a = np.array([1.0], dtype=np.float32)
+        >>> b = np.array(1.0, dtype=np.float64)
+        >>> np.add(a, b) # The spec says this should be float64
+        array([2.], dtype=float32)
+
+        To fix this, we add a dimension to the 0-dimension array before passing it
+        through. This works because a dimension would be added anyway from
+        broadcasting, so the resulting shape is the same, but this prevents NumPy
+        from not promoting the dtype.
+        """
+        if x1.shape == () and x2.shape != ():
+            # The _array[None] workaround was chosen because it is relatively
+            # performant. broadcast_to(x1._array, x2.shape) is much slower. We
+            # could also manually type promote x2, but that is more complicated
+            # and about the same performance as this.
+            x1 = ndarray._new(x1._array[None])
+        elif x2.shape == () and x1.shape != ():
+            x2 = ndarray._new(x2._array[None])
+        return (x1, x2)
+
+    # Everything below this line is required by the spec.
+
+    def __abs__(self: array, /) -> array:
+        """
+        Performs the operation __abs__.
+        """
+        res = self._array.__abs__()
+        return self.__class__._new(res)
+
+    @np.errstate(all='ignore')
+    def __add__(self: array, other: Union[int, float, array], /) -> array:
+        """
+        Performs the operation __add__.
+        """
+        if isinstance(other, (int, float, bool)):
+            other = self._promote_scalar(other)
+        self, other = self._normalize_two_args(self, other)
+        res = self._array.__add__(other._array)
+        return self.__class__._new(res)
+
+    def __and__(self: array, other: Union[int, bool, array], /) -> array:
+        """
+        Performs the operation __and__.
+        """
+        if isinstance(other, (int, float, bool)):
+            other = self._promote_scalar(other)
+        self, other = self._normalize_two_args(self, other)
+        res = self._array.__and__(other._array)
+        return self.__class__._new(res)
+
+    def __array_namespace__(self: array, /, *, api_version: Optional[str] = None) -> object:
+        if api_version is not None:
+            raise ValueError("Unrecognized array API version")
+        from numpy import _array_api
+        return _array_api
+
+    def __bool__(self: array, /) -> bool:
+        """
+        Performs the operation __bool__.
+        """
+        # Note: This is an error here.
+        if self._array.shape != ():
+            raise TypeError("bool is only allowed on arrays with shape ()")
+        res = self._array.__bool__()
+        return res
+
+    def __dlpack__(self: array, /, *, stream: Optional[Union[int, Any]] = None) -> PyCapsule:
```

Comment:
```suggestion
    def __dlpack__(self: array, /, *, stream: Optional[Any] = None) -> PyCapsule:
```
There is a bit of typing subtlety here due to the fact that protocol parameters (and protocol-esque APIs in general) are contravariant, _i.e._ they denote what a concrete implementation must support at minimum.
With `None | int | Any` you're thus mandating that implementations _must_ support both `int` and `None`, a mandate that does not seem to be correct based on prior discussion in https://github.com/numpy/numpy/pull/19083.

Reply:
It's a bit tricky to express (hence why I'm suggesting `None | Any`).
Namely, `None` is the absolute minimum bound here, but user implementation are allowed to add
additional lower bounds (_e.g._ froom `None` to `int | None`).

Reply:
If I understand it correctly, for this implementation this should just be `None`. 

I actually don't understand why `Union[int, Any]` even does anything at all. Wouldn't the union of anything with `Any` just be the same as `Any`?

Reply:
> If I understand it correctly, for this implementation this should just be `None`.

Ah, you're right actually. 
I missed for a moment that this NEP 47 wrapper is exclusively for `np.ndarray` 
(so no need to worry about other tensor types that _could_ support `int`-based parameters).

Reply:
Well I've also been trying to update the type hints in the spec itself concurrently with my updates on this PR based on your feedback, so this is still a relevant question. Should we change the type of `stream` in the spec from `Optional[int, Any]` to just `Optional[Any]`? I admit I'm still a bit confused by `Any`. The [Python docs](https://docs.python.org/3/library/typing.html#the-any-type) state "A special kind of type is Any. A static type checker will treat every type as being compatible with Any and Any as being compatible with every type.", so wouldn't `Union[int, Any]` or indeed `Optional[Any]` just be equivalent to just `Any` as far as type checking is concerned? It seems to me that the only point of `Optional[int, Any]` would be to document that `None` and `int` have special meaning, but based on what you've stated here, this is perhaps just me misunderstanding how Any works with respect to covariance and contravariance. 

I'm also a bit confused by the `Any` type for this specific `__dlpack__` signature. Is it really Any, or should it be an unspecified `<Stream>` type, similar to what is used for `device`? Perhaps we should move this discussion over to https://github.com/data-apis/array-api/issues/143.